Improve images doc and add new image formats (#404)
springernature/frontend-playbook: performance/images.md

* **Content images**. These are unique to a page, or a handful of pages. They will usually be provided from a database or a content management system.

Most best practices described in this document apply to both categories unless otherwise indicated.

## Picking the right image format

* Use [SVG](https://en.wikipedia.org/wiki/Scalable_Vector_Graphics) for icons, logos, and other images with sharply defined edges.
* Use [JPEG](https://en.wikipedia.org/wiki/JPEG) for still images, photographs, and in any other cases where using a PNG would result in an image too large to be delivered over the network effectively.
* Use [PNG](https://en.wikipedia.org/wiki/Portable_Network_Graphics) for illustrations, graphs, flat-colour graphics, images containing transparency (which JPEG doesn't support), and images containing text.
* Use [MP4](https://en.wikipedia.org/wiki/MPEG-4_Part_14) for any kind of animated images.

> * Do not use [GIFs](https://en.wikipedia.org/wiki/GIF). For animated images, the file size of GIF files, usually an order of magnitude bigger than MP4s of similar quality, may be a barrier preventing users from accessing our content, specially in markets where slower connections and low performing devices are more common. For static images with transparency, an 8-bit PNG can be used to deliver an indexed colour image, with binary transparency, [at a smaller size than a GIF](https://helpx.adobe.com/photoshop-elements/using/optimizing-images-gif-or-png.html).

Reviewer: What about images without transparency … PNGs still smaller?

Author: It should, yes. The reason why the text only mentions the transparency use case is because GIFs have been traditionally used for this purpose. But from the Adobe link mentioned:

> The PNG‑8 format uses more advanced compression schemes than GIF does, and a PNG‑8 file can be 10% to 30% smaller than a GIF file of the same image, depending on the image’s color patterns.

This is independent from the image using transparency or not.

> * Do not use [GIFs](https://en.wikipedia.org/wiki/GIF). For animated images, the file size of GIF files, usually an order of magnitude bigger than MP4s of similar quality, may be a barrier preventing users from accessing our content, specially in markets where slower connections and low performing devices are more common. For static images with transparency, an 8-bit PNG can be used to deliver an indexed colour image, with binary transparency, [at a smaller size than a GIF](https://helpx.adobe.com/photoshop-elements/using/optimizing-images-gif-or-png.html).

Reviewer 2: ```suggestion
* Do not use [GIFs](https://en.wikipedia.org/wiki/GIF). The file size of animated GIF images is usually an order of magnitude bigger than MP4s of similar quality. They can be a barrier that prevents users from accessing our content, especially in markets where slower connections and low performing devices are more common. For static images with transparency, an 8-bit PNG can be used to deliver an indexed colour image, with binary transparency, [at a smaller size than a GIF](https://helpx.adobe.com/photoshop-elements/using/optimizing-images-gif-or-png.html).
```

Turns one long long ~~man~~ sentence into two shorter ones

> A number of more modern image formats can be also used to deliver images:

Reviewer 3: ```suggestion
A number of more modern image formats can also be used to deliver images:
```

> * [WEBP](https://en.wikipedia.org/wiki/WebP) is a replacement for JPEG, PNG, and GIFs, but improvements to encoders like mozjpeg means that WEBP file sizes are only marginally smaller than JPEG. It's supported by almost all major browsers ([Safari supports it partially](https://caniuse.com/webp)) and can be used by providing another better supported format as a fallback.

Reviewer: ```suggestion
* [WEBP](https://en.wikipedia.org/wiki/WebP) is a replacement for JPEG, PNG, and GIFs, but improvements to encoders like mozjpeg means that WEBP file sizes are only marginally smaller than JPEG. It's supported by almost all major browsers ([Safari supports it partially](https://caniuse.com/webp)) and can be used only as an enhancement to another better supported format.
```
Using the language of progressive enhancement rather than graceful degradation.

> * [AVIF](https://en.wikipedia.org/wiki/AVIF) is suitable for static and animated images, and supports both lossy and lossless compression with better compression efficiency and better detail preservation than other formats. It's [supported by Chrome and Firefox](https://caniuse.com/avif), and can be used by providing another better supported format as a fallback. Compression (encoding) of images can be significantly slower than with other formats.

Reviewer: ```suggestion
* [AVIF](https://en.wikipedia.org/wiki/AVIF) is suitable for static and animated images, and supports both lossy and lossless compression with better compression efficiency and better detail preservation than other formats. It's [supported by Chrome and Firefox](https://caniuse.com/avif), and can be used only as an enhancement to another better supported format. Compression (encoding) of images can be significantly slower than with other formats.
```
Using the language of progressive enhancement  rather than graceful degradation.

* [JPEG-XL](https://en.wikipedia.org/wiki/JPEG_XL) is in the process of becoming a standard and supports both lossy and lossless compression. It's [more efficient than other formats while being usually faster](https://cloudinary.com/blog/how_jpeg_xl_compares_to_other_image_codecs

> ) both when compressing and decompressing than JPEG. [No browser officially supports it](https://caniuse.com/jpegxl) at the time of writing.

Author: 👍

* Use progressive encoding. Progressive JPEGs provide a better user experience and, most of the time, are smaller than baseline. See this [yeoman issue](https://github.com/yeoman/yeoman/issues/810) for further details.
* Use [4:2:0](https://en.wikipedia.org/wiki/Chroma_subsampling#4:2:0) [chroma subsampling](https://en.wikipedia.org/wiki/Chroma_subsampling) whenever possible. This encodes the colour information at half the resolution of the luma, which may result in a perceptible loss of colour accuracy. This will usually generate much smaller images, so we suggest to create both of them and compare them for any loss of quality. For tools that can help you with this, see: [Comparing the quality of images](#comparing-the-quality-of-images), below.
* Use encoders like [Mozilla's mozjpeg](https://github.com/mozilla/mozjpeg) that can create images with the same visual quality as other encoders but a much smaller file size.

### PNG images

Always use the maximum compression level (9).

## Optimisation of images

> Images can frequently make up a significant amount of a page's size, so it's important to optimise them before they are delivered to the users in order to ensure that our sites stay performant and we're not preventing users from accessing the site due to the increased page weight. Common optimisations include:

Reviewer 3: ```suggestion
Images can contribute significantly to a page's size, so it's important to optimise them before they are delivered to the users. This helps ensure that our sites stay performant and we're not preventing anyone from accessing them due to increased page weight. Common optimisations include:
```

If an image is still too big after following all these optimisations, other options can be used that will require editing the image. These can be used to generate images with a much smaller size than otherwise would be possible:

* Reducing noise or eliminating noise and other complexity from the image, which makes it easier to compress.
* Blurring unimportant areas will also make the image easier to compress.
* Finally, decreasing the export quality may be required if the image is still too big.

## Serving images

We work under the assumption that images are encoded only once but decoded many times. Therefore, increased encoding times, while undesirable, are usually secondary to other concerns like making the image easier to decompress, or having a smaller file size.

When optimising non-content images that are going to be committed to a repo, make sure you always use the maximum settings if you're using lossless optimisations.

> For content images, we encourage using an on-premise or cloud-based image server in order to:

Reviewer 2: should we mention Spindoctor here? or is it too platform-specific?

Reviewer 4: I wanted to mention the quality settings from Spindoctor 😬 but this could also be something to reference about the quality setting.

From the Spindoctor README:
>By default, PNGs will be output at compression quality of 95, and JPEGs at 90 (for lossless sources such as PNG or TIFF) or 80 (for lossy sources).

Those seem quite high? This is just a feeling and I don't know how those original settings came about. I suppose with photographic images you could get away with lower but it may be more noticeable against figure images with graphs.

Author: Spindoctor is not public or open source, so I thought it might be confusing for external readers to see a product mentioned that is not publicly available. I also don't feel like we should endorse any products, which is why I remove the reference to specific services.

> Those seem quite high? This is just a feeling and I don't know how those original settings came about. I suppose with photographic images you could get away with lower but it may be more noticeable against figure images with graphs.

Correct. Defaults were adjusted in April 2020. Happy to discuss specifics in Slack.

* Ensure that all images are optimised automatically.
* Ensure that the appropriate cache headers are always set.

Additionally, future improvements to these tools will result in improvements to the images being served automatically, without the images having to be manually re-encoded.

## Comparing the quality of images

> When comparing the output of two different image compression formats, algorithms, or quality settings, [tools](https://github.com/kornelski/dssim) that calculate the similarity between images with algorithms like [SSIM](http://en.wikipedia.org/wiki/Structural_similarity) and DSSIM can be used to get an objective idea of the trade-off between quality and byte size.

Reviewer 2: ```suggestion
When comparing the output of two different image compression formats, algorithms, or quality settings, [tools that calculate the similarity between images](https://github.com/kornelski/dssim) with algorithms like [SSIM](http://en.wikipedia.org/wiki/Structural_similarity) and DSSIM can be used to get an objective idea of the trade-off between quality and byte size.
```
